matcalc/elasticity.py

```diff
@@ -53,33 +66,75 @@ def calc(self, structure: Structure) -> dict:
         }
         """
         if self.relax_structure:
-            rcalc = RelaxCalc(self.calculator, fmax=self.fmax)
-            structure = rcalc.calc(structure)["final_structure"]
+            relax_calc = RelaxCalc(self.calculator, fmax=self.fmax)
+            structure = relax_calc.calc(structure)["final_structure"]
 
-        adaptor = AseAtomsAdaptor()
         deformed_structure_set = DeformedStructureSet(
             structure,
-            [self.norm_strains],
-            [self.shear_strains],
+            self.norm_strains,
+            self.shear_strains,
         )
         stresses = []
         for deformed_structure in deformed_structure_set:
-            atoms = adaptor.get_atoms(deformed_structure)
+            atoms = deformed_structure.to_ase_atoms()
             atoms.calc = self.calculator
             stresses.append(atoms.get_stress(voigt=False))
 
         strains = [Strain.from_deformation(deformation) for deformation in deformed_structure_set.deformations]
-        atoms = adaptor.get_atoms(structure)
+        atoms = structure.to_ase_atoms()
         atoms.calc = self.calculator
-        elastic_tensor = ElasticTensor.from_independent_strains(
-            strains, stresses, eq_stress=atoms.get_stress(voigt=False)
+        elastic_tensor, residuals_sum = self._elastic_tensor_from_strains(
+            strains,
+            stresses,
+            eq_stress=atoms.get_stress(voigt=False) if self.use_equilibrium else None,
         )
         return {
             "elastic_tensor": elastic_tensor,
             "shear_modulus_vrh": elastic_tensor.g_vrh,
             "bulk_modulus_vrh": elastic_tensor.k_vrh,
             "youngs_modulus": elastic_tensor.y_mod,
+            "residuals_sum": residuals_sum,
+            "structure": structure,
         }
+
+    def _elastic_tensor_from_strains(
+        self,
+        strains,
+        stresses,
+        eq_stress=None,
+        tol: float = 1e-7,
+    ):
+        """
+        Slightly modified version of Pymatgen function
+        pymatgen.analysis.elasticity.elastic.ElasticTensor.from_independent_strains;
+        this is to give option to discard eq_stress,
+        which (if the structure is relaxed) tends to sometimes be
+        much lower than neighboring points.
+        Also has option to return the sum of the squares of the residuals
+        for all of the linear fits done to compute the entries of the tensor.
+        """
+        strain_states = [tuple(ss) for ss in np.eye(6)]
+        ss_dict = get_strain_state_dict(strains, stresses, eq_stress=eq_stress, add_eq=self.use_equilibrium)
+        if set(strain_states) >= set(ss_dict):
+            raise ValueError(f"Missing independent strain states: {set(strain_states) - set(ss_dict)}")
+        if set(strain_states) <= set(ss_dict):
+            warnings.warn(
+                "Extra strain states in strain-stress pairs are neglected in independent strain fitting", stacklevel=1
+            )
```

Comment:
This and a few other lines need test coverage. See the coverage report comments. I started `test_elastic_calc_invalid_states`, feel free to add to it until we're back at 100% coverage.